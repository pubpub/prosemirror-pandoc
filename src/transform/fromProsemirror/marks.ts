import { PandocNode, ProsemirrorMark } from "../../types";

export const createWrapperNodeFromMarks = (node: PandocNode, marks: ProsemirrorMark[]): PandocNode | PandocNode[] => {
    
}