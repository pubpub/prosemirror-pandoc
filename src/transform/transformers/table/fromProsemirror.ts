import {
    Block,
    Caption,
    Cell,
    ColSpec,
    ProsemirrorNode,
    Row,
    Table,
} from "types";
import { FromProsemirrorTransformContext } from "transform/types";
import { createAttr } from "transform/util";

const getDefaultColSpec = (): ColSpec => ({
    type: "ColSpec",
    alignment: "AlignDefault",
    defaultWidth: true,
});

const getDefaultCaption = (): Caption => ({
    type: "Caption",
    content: [],
});

const getColumnCountFromRow = (row: ProsemirrorNode<"table_row">) => {
    return row.content
        .map((cell) => cell.attrs?.colspan ?? 1)
        .map((attr) => Number(attr))
        .reduce((a, b) => a + b);
};

const getColSpecsForTable = (
    header: ProsemirrorNode<"table_row">,
    context: FromProsemirrorTransformContext
) => {
    const { prosemirrorDocWidth } = context;
    const columnCount = getColumnCountFromRow(header);
    const colSpecs: ColSpec[] = new Array(columnCount)
        .fill(0)
        .map(() => getDefaultColSpec());
    header.content.forEach((cell, index) => {
        if ("colwidth" in cell.attrs && cell.attrs.colwidth) {
            const colWidth = cell.attrs.colwidth as (number | null)[];
            colWidth.forEach((width, cellIndex) => {
                const realColumnIndex = index + cellIndex;
                if (typeof width === "number") {
                    colSpecs[realColumnIndex] = {
                        type: "ColSpec",
                        alignment: "AlignDefault",
                        width: width / prosemirrorDocWidth,
                    };
                }
            });
        }
    });
    return colSpecs;
};

const transformCell = (
    cell: ProsemirrorNode<"table_cell" | "table_header">,
    context: FromProsemirrorTransformContext
): Cell => {
    const { transform } = context;
    const { colspan = 1, rowspan = 1 } = cell.attrs;
    return {
        type: "Cell",
        attr: createAttr(),
        alignment: "AlignDefault",
        rowSpan: Number(rowspan),
        colSpan: Number(colspan),
        content: transform(cell.content).asArray() as Block[],
    };
};

const transformRow = (
    row: ProsemirrorNode<"table_row">,
    context: FromProsemirrorTransformContext
): Row => {
    const cells = row.content as ProsemirrorNode<
        "table_cell" | "table_header"
    >[];
    return {
        type: "Row",
        attr: createAttr(),
        // Table rows may be devoid of content, for example phantom rows that
        // are automatically added to satisfy a row where each element has
        // rowspan >1
        cells: cells?.map((cell) => transformCell(cell, context)) ?? [] ,
    };
};

export const prosemirrorTableTransformer = (
    table: ProsemirrorNode<"table">,
    context: FromProsemirrorTransformContext
): Table => {
    const [header, ...body] = table.content as ProsemirrorNode<"table_row">[];
    return {
        type: "Table",
        attr: createAttr("id" in table.attrs ? String(table.attrs.id) : ""),
        caption: getDefaultCaption(),
        colSpecs: getColSpecsForTable(header, context),
        head: {
            type: "TableHead",
            attr: createAttr(),
            rows: [transformRow(header, context)],
        },
        bodies: [
            {
                type: "TableBody",
                attr: createAttr(),
                rowHeadColumns: 1,
                headRows: [],
                bodyRows: body.map((row) => transformRow(row, context)),
            },
        ],
        foot: {
            type: "TableFoot",
            attr: createAttr(),
            rows: [],
        },
    };
};

prosemirrorTableTransformer.assertCapturedProsemirrorNodes = [
    "table_row",
    "table_cell",
    "table_header",
];
