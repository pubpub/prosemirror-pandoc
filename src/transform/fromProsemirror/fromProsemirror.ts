import { ProsemirrorNode, PandocNode } from "types";

import { asArray, flatten, makeCounter } from "transform/util";
import { RuleSet } from "transform/ruleset";
import { Fluent, fluent } from "transform/fluent";
import {
    FromProsemirrorTransformConfig,
    FromProsemirrorTransformContext,
} from "transform/types";

import { createWrapperNodeFromMarks, splitNodesByMarks } from "./marks";

const fromProsemirrorInner = (
    elementOrArray: ProsemirrorNode | ProsemirrorNode[],
    context: FromProsemirrorTransformContext
): Fluent<PandocNode> => {
    if (!elementOrArray) {
        return fluent([]);
    }
    const { ruleset } = context;
    const nodesAndAssociatedMarks = splitNodesByMarks(asArray(elementOrArray));
    const transformed: PandocNode[] = [];
    for (const { nodes, marks } of nodesAndAssociatedMarks) {
        let ptr = 0;
        const innerTransformed = [];
        while (ptr < nodes.length) {
            const remaining = nodes.slice(ptr);
            const { rule, acceptedCount } =
                ruleset.matchProsemirrorNodes(remaining);
            const addition: PandocNode[] = flatten(
                rule.acceptsMultiple
                    ? rule.transformer(
                          nodes.slice(ptr, ptr + acceptedCount),
                          context
                      )
                    : rule.transformer(nodes[ptr], context)
            );
            innerTransformed.push(...addition);
            ptr += acceptedCount;
        }
        const maybeWrappedNodes = createWrapperNodeFromMarks(
            innerTransformed,
            marks,
            context
        );
        transformed.push(...asArray(maybeWrappedNodes));
    }
    return fluent(transformed);
};

export const fromProsemirror = (
    elementOrArray: ProsemirrorNode | ProsemirrorNode[],
    ruleset: RuleSet<any>,
    config: Partial<FromProsemirrorTransformConfig> = {}
): Fluent<PandocNode> => {
    const { resource = (x) => x, prosemirrorDocWidth = 1000 } = config;
    const context: FromProsemirrorTransformContext = {
        ruleset,
        resource,
        prosemirrorDocWidth,
        count: makeCounter(),
        transform: (element) => fromProsemirrorInner(element, context),
    };
    return context.transform(elementOrArray);
};
