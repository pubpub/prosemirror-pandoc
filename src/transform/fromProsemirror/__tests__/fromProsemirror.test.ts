import { rules } from "example/rules";
import { createAttr } from "transform/util";

import { fromProsemirror } from "../fromProsemirror";

describe("fromProsemirror", () => {
    it("transforms a doc into a Doc", () => {
        expect(
            fromProsemirror(
                {
                    type: "doc",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Hello dad!" }],
                        },
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Hello mom!" }],
                        },
                        {
                            type: "paragraph",
                            content: [
                                {
                                    type: "text",
                                    text: "Ch-ch-ch-ch-ch-cherry bomb!",
                                },
                            ],
                        },
                    ],
                },
                rules
            ).asNode()
        ).toEqual({
            type: "Doc",
            meta: {},
            blocks: [
                {
                    type: "Para",
                    content: [
                        { type: "Str", content: "Hello" },
                        { type: "Space" },
                        { type: "Str", content: "dad!" },
                    ],
                },
                {
                    type: "Para",
                    content: [
                        { type: "Str", content: "Hello" },
                        { type: "Space" },
                        { type: "Str", content: "mom!" },
                    ],
                },
                {
                    type: "Para",
                    content: [
                        {
                            type: "Str",
                            content: "Ch-ch-ch-ch-ch-cherry",
                        },
                        { type: "Space" },
                        { type: "Str", content: "bomb!" },
                    ],
                },
            ],
        });
    });

    it("transforms a simple string", () => {
        expect(
            fromProsemirror(
                { type: "text", text: "Hello world!" },
                rules
            ).asArray()
        ).toEqual([
            { type: "Str", content: "Hello" },
            { type: "Space" },
            { type: "Str", content: "world!" },
        ]);
    });

    it("transforms a string with trailing spaces", () => {
        expect(
            fromProsemirror(
                { type: "text", text: "Hello world " },
                rules
            ).asArray()
        ).toEqual([
            { type: "Str", content: "Hello" },
            { type: "Space" },
            { type: "Str", content: "world" },
            { type: "Space" },
        ]);
    });

    it("performs a simple transformation from Prosemirror marks to nodes", () => {
        expect(
            fromProsemirror(
                {
                    type: "paragraph",
                    content: [
                        {
                            type: "text",
                            marks: [{ type: "strong" }],
                            text: "Hello world!",
                        },
                    ],
                },
                rules
            ).asNode()
        ).toEqual({
            type: "Para",
            content: [
                {
                    type: "Strong",
                    content: [
                        { type: "Str", content: "Hello" },
                        { type: "Space" },
                        { type: "Str", content: "world!" },
                    ],
                },
            ],
        });
    });

    it("transforms multiple marks into nested Pandoc nodes", () => {
        expect(
            fromProsemirror(
                {
                    type: "paragraph",
                    content: [
                        {
                            type: "text",
                            marks: [{ type: "strong" }, { type: "em" }],
                            text: "Hello world!",
                        },
                        {
                            type: "text",
                            marks: [{ type: "strong" }, { type: "em" }],
                            text: "Hello again!!",
                        },
                    ],
                },
                rules
            ).asNode()
        ).toEqual({
            type: "Para",
            content: [
                {
                    type: "Emph",
                    content: [
                        {
                            type: "Strong",
                            content: [
                                { type: "Str", content: "Hello" },
                                { type: "Space" },
                                { type: "Str", content: "world!" },
                                { type: "Str", content: "Hello" },
                                { type: "Space" },
                                { type: "Str", content: "again!!" },
                            ],
                        },
                    ],
                },
            ],
        });
    });

    it("transforms a code_block into a CodeBlock", () => {
        expect(
            fromProsemirror(
                {
                    type: "code_block",
                    content: [{ type: "text", text: "hello_world()" }],
                },
                rules
            ).asNode()
        ).toEqual({
            type: "CodeBlock",
            content: "hello_world()",
            attr: createAttr(""),
        });
    });

    it("transforms a blockquote into a BlockQuote", () => {
        expect(
            fromProsemirror(
                {
                    type: "blockquote",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Lorem Ipsum?" }],
                        },
                        {
                            type: "paragraph",
                            content: [
                                {
                                    type: "text",
                                    text: "More like Borem Ipsum!",
                                },
                            ],
                        },
                    ],
                },
                rules
            ).asNode()
        ).toEqual({
            type: "BlockQuote",
            content: [
                {
                    type: "Para",
                    content: [
                        { type: "Str", content: "Lorem" },
                        { type: "Space" },
                        { type: "Str", content: "Ipsum?" },
                    ],
                },
                {
                    type: "Para",
                    content: [
                        { type: "Str", content: "More" },
                        { type: "Space" },
                        { type: "Str", content: "like" },
                        { type: "Space" },
                        { type: "Str", content: "Borem" },
                        { type: "Space" },
                        { type: "Str", content: "Ipsum!" },
                    ],
                },
            ],
        });
    });

    it("transforms a heading into a header", () => {
        expect(
            fromProsemirror(
                {
                    type: "heading",
                    attrs: {
                        level: 1,
                        id: "lesson-one",
                    },
                    content: [{ type: "text", text: "Lesson One" }],
                },
                rules
            ).asNode()
        ).toEqual({
            type: "Header",
            attr: createAttr("lesson-one", [], {}),
            level: 1,
            content: [
                { type: "Str", content: "Lesson" },
                { type: "Space" },
                { type: "Str", content: "One" },
            ],
        });
    });

    it("transforms a horizontal_rule into a HorizontalRule", () => {
        expect(
            fromProsemirror(
                {
                    type: "horizontal_rule",
                },
                rules
            ).asNode()
        ).toEqual({ type: "HorizontalRule" });
    });

    it("transforms a bullet_list into a BulletList", () => {
        const input = {
            type: "bullet_list",
            attrs: {},
            content: [
                {
                    type: "list_item",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Item One" }],
                        },
                    ],
                },
                {
                    type: "list_item",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Item Two" }],
                        },
                    ],
                },
                {
                    type: "list_item",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Item Three" }],
                        },
                    ],
                },
            ],
        };

        const output = {
            type: "BulletList",
            content: [
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Item" },
                            { type: "Space" },
                            { type: "Str", content: "One" },
                        ],
                    },
                ],
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Item" },
                            { type: "Space" },
                            { type: "Str", content: "Two" },
                        ],
                    },
                ],
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Item" },
                            { type: "Space" },
                            { type: "Str", content: "Three" },
                        ],
                    },
                ],
            ],
        };

        expect(fromProsemirror(input, rules).asNode()).toEqual(output);
    });

    it("transforms an ordered_list into an OrderedList", () => {
        const input = {
            type: "ordered_list",
            attrs: {
                order: 2,
            },
            content: [
                {
                    type: "list_item",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Item One" }],
                        },
                    ],
                },
                {
                    type: "list_item",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Item Two" }],
                        },
                    ],
                },
                {
                    type: "list_item",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Item Three" }],
                        },
                    ],
                },
            ],
        };

        const output = {
            type: "OrderedList",
            listAttributes: {
                startNumber: 2,
                listNumberDelim: "DefaultDelim",
                listNumberStyle: "DefaultStyle",
            },
            content: [
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Item" },
                            { type: "Space" },
                            { type: "Str", content: "One" },
                        ],
                    },
                ],
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Item" },
                            { type: "Space" },
                            { type: "Str", content: "Two" },
                        ],
                    },
                ],
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Item" },
                            { type: "Space" },
                            { type: "Str", content: "Three" },
                        ],
                    },
                ],
            ],
        };

        expect(fromProsemirror(input, rules).asNode()).toEqual(output);
    });

    it("transforms an equation into a Math (mathType=InlineMath)", () => {
        expect(
            fromProsemirror(
                {
                    type: "equation",
                    attrs: { value: "e^{i\\pi} = -1" },
                },
                rules
            ).asNode()
        ).toEqual({
            type: "Math",
            mathType: "InlineMath",
            content: "e^{i\\pi} = -1",
        });
    });

    it("transforms a block_equation into a Plain wrapping a Math (mathType=DisplayMath)", () => {
        expect(
            fromProsemirror(
                {
                    type: "block_equation",
                    attrs: { value: "e^{i\\pi} = -1" },
                },
                rules
            ).asNode()
        ).toEqual({
            type: "Plain",
            content: [
                {
                    type: "Math",
                    mathType: "DisplayMath",
                    content: "e^{i\\pi} = -1",
                },
            ],
        });
    });

    it("transforms links", () => {
        const link = {
            type: "text",
            marks: [
                {
                    type: "link",
                    attrs: { href: "https://example.com", title: "Hi!" },
                },
            ],
            text: "Hey!",
        };
        expect(fromProsemirror(link, rules).asNode()).toEqual({
            type: "Link",
            attr: createAttr(""),
            target: {
                url: "https://example.com",
                title: "Hi!",
            },
            content: [{ type: "Str", content: "Hey!" }],
        });
    });

    it("handles an image with no alt text", () => {
        expect(
            fromProsemirror(
                {
                    type: "image",
                    attrs: {
                        url: "https://pubpub.org/logo.png",
                    },
                },
                rules
            ).asNode()
        ).toEqual({
            type: "Plain",
            content: [
                {
                    type: "Image",
                    target: {
                        title: "",
                        url: "https://pubpub.org/logo.png",
                    },
                    attr: createAttr(""),
                    content: [],
                },
            ],
        });
    });

    it("handles an image with alt text", () => {
        expect(
            fromProsemirror(
                {
                    type: "image",
                    attrs: {
                        url: "https://pubpub.org/logo.png",
                        altText: "Very cool.",
                    },
                },
                rules
            ).asNode()
        ).toEqual({
            type: "Plain",
            content: [
                {
                    type: "Image",
                    target: {
                        title: "",
                        url: "https://pubpub.org/logo.png",
                    },
                    attr: createAttr(""),
                    content: [
                        { type: "Str", content: "Very" },
                        { type: "Space" },
                        { type: "Str", content: "cool." },
                    ],
                },
            ],
        });
    });

    it("handles an image with a caption", () => {
        expect(
            fromProsemirror(
                {
                    type: "image",
                    attrs: {
                        url: "https://pubpub.org/logo.png",
                        altText: "Very cool.",
                        caption: "<b>Even</b> cooler.",
                    },
                },
                rules
            ).asArray()
        ).toEqual([
            {
                type: "Plain",
                content: [
                    {
                        type: "Image",
                        target: {
                            title: "",
                            url: "https://pubpub.org/logo.png",
                        },
                        attr: createAttr(""),
                        content: [
                            { type: "Str", content: "Very" },
                            { type: "Space" },
                            { type: "Str", content: "cool." },
                        ],
                    },
                ],
            },
            {
                type: "Plain",
                content: [
                    {
                        type: "Strong",
                        content: [{ type: "Str", content: "Even" }],
                    },
                    { type: "Space" },
                    { type: "Str", content: "cooler." },
                ],
            },
        ]);
    });

    it("converts a table", () => {
        expect(
            fromProsemirror(
                {
                    type: "table",
                    attrs: {},
                    content: [
                        {
                            type: "table_row",
                            content: [
                                {
                                    type: "table_header",
                                    attrs: {
                                        colspan: 1,
                                        rowspan: 1,
                                        colwidth: [134],
                                    },
                                    content: [
                                        {
                                            type: "paragraph",
                                            attrs: {
                                                id: "nnbagxoadui",
                                                class: null,
                                                textAlign: null,
                                            },
                                        },
                                    ],
                                },
                                {
                                    type: "table_cell",
                                    attrs: {
                                        colspan: 1,
                                        rowspan: 1,
                                        colwidth: [333],
                                    },
                                    content: [
                                        {
                                            type: "paragraph",
                                            attrs: {
                                                id: "nlqg7398ode",
                                                class: null,
                                                textAlign: null,
                                            },
                                            content: [
                                                {
                                                    type: "text",
                                                    text: "He-Man",
                                                },
                                            ],
                                        },
                                    ],
                                },
                                {
                                    type: "table_cell",
                                    attrs: {
                                        colspan: 1,
                                        rowspan: 1,
                                        colwidth: [269],
                                    },
                                    content: [
                                        {
                                            type: "paragraph",
                                            attrs: {
                                                id: "n29gzcvua5p",
                                                class: null,
                                                textAlign: null,
                                            },
                                            content: [
                                                {
                                                    type: "text",
                                                    text: "Skeletor",
                                                },
                                            ],
                                        },
                                    ],
                                },
                            ],
                        },
                        {
                            type: "table_row",
                            content: [
                                {
                                    type: "table_header",
                                    attrs: {
                                        colspan: 1,
                                        rowspan: 1,
                                        colwidth: [134],
                                    },
                                    content: [
                                        {
                                            type: "paragraph",
                                            attrs: {
                                                id: "nvt96w89tmm",
                                                class: null,
                                                textAlign: null,
                                            },
                                            content: [
                                                { type: "text", text: "Role" },
                                            ],
                                        },
                                    ],
                                },
                                {
                                    type: "table_cell",
                                    attrs: {
                                        colspan: 1,
                                        rowspan: 1,
                                        colwidth: [333],
                                    },
                                    content: [
                                        {
                                            type: "paragraph",
                                            attrs: {
                                                id: "nv0eioahc5d",
                                                class: null,
                                                textAlign: null,
                                            },
                                            content: [
                                                { type: "text", text: "Hero" },
                                            ],
                                        },
                                    ],
                                },
                                {
                                    type: "table_cell",
                                    attrs: {
                                        colspan: 1,
                                        rowspan: 1,
                                        colwidth: [269],
                                    },
                                    content: [
                                        {
                                            type: "paragraph",
                                            attrs: {
                                                id: "ndfs47q3sp4",
                                                class: null,
                                                textAlign: null,
                                            },
                                            content: [
                                                {
                                                    type: "text",
                                                    text: "Villain",
                                                },
                                            ],
                                        },
                                    ],
                                },
                            ],
                        },
                        {
                            type: "table_row",
                            content: [
                                {
                                    type: "table_header",
                                    attrs: {
                                        colspan: 1,
                                        rowspan: 1,
                                        colwidth: [134],
                                    },
                                    content: [
                                        {
                                            type: "paragraph",
                                            attrs: {
                                                id: "njy6g19emna",
                                                class: null,
                                                textAlign: null,
                                            },
                                            content: [
                                                {
                                                    type: "text",
                                                    text: "Weapon",
                                                },
                                            ],
                                        },
                                    ],
                                },
                                {
                                    type: "table_cell",
                                    attrs: {
                                        colspan: 1,
                                        rowspan: 1,
                                        colwidth: [333],
                                    },
                                    content: [
                                        {
                                            type: "paragraph",
                                            attrs: {
                                                id: "nhp0euqi7j9",
                                                class: null,
                                                textAlign: null,
                                            },
                                            content: [
                                                {
                                                    type: "text",
                                                    text: "Power Sword",
                                                },
                                            ],
                                        },
                                    ],
                                },
                                {
                                    type: "table_cell",
                                    attrs: {
                                        colspan: 1,
                                        rowspan: 1,
                                        colwidth: [269],
                                    },
                                    content: [
                                        {
                                            type: "paragraph",
                                            attrs: {
                                                id: "nfif0x4mhfd",
                                                class: null,
                                                textAlign: null,
                                            },
                                            content: [
                                                {
                                                    type: "text",
                                                    text: "Havoc Staff",
                                                },
                                            ],
                                        },
                                    ],
                                },
                            ],
                        },
                        {
                            type: "table_row",
                            content: [
                                {
                                    type: "table_header",
                                    attrs: {
                                        colspan: 1,
                                        rowspan: 1,
                                        colwidth: [134],
                                    },
                                    content: [
                                        {
                                            type: "paragraph",
                                            attrs: {
                                                id: "nm0w3am87hj",
                                                class: null,
                                                textAlign: null,
                                            },
                                            content: [
                                                {
                                                    type: "text",
                                                    text: "Dark secret",
                                                },
                                            ],
                                        },
                                    ],
                                },
                                {
                                    type: "table_cell",
                                    attrs: {
                                        colspan: 1,
                                        rowspan: 1,
                                        colwidth: [333],
                                    },
                                    content: [
                                        {
                                            type: "paragraph",
                                            attrs: {
                                                id: "ne2rjff3y09",
                                                class: null,
                                                textAlign: null,
                                            },
                                            content: [
                                                {
                                                    type: "text",
                                                    text: "Expert florist",
                                                },
                                            ],
                                        },
                                    ],
                                },
                                {
                                    type: "table_cell",
                                    attrs: {
                                        colspan: 1,
                                        rowspan: 1,
                                        colwidth: [269],
                                    },
                                    content: [
                                        {
                                            type: "paragraph",
                                            attrs: {
                                                id: "nj0u47p9psl",
                                                class: null,
                                                textAlign: null,
                                            },
                                            content: [
                                                {
                                                    type: "text",
                                                    text: "Cries at romcoms",
                                                },
                                            ],
                                        },
                                    ],
                                },
                            ],
                        },
                        {
                            type: "table_row",
                            content: [
                                {
                                    type: "table_cell",
                                    attrs: {
                                        colspan: 3,
                                        rowspan: 1,
                                        colwidth: [134, 333, 269],
                                    },
                                    content: [
                                        {
                                            type: "paragraph",
                                            attrs: {
                                                id: "n6nffgb8rbn",
                                                class: null,
                                                textAlign: "center",
                                            },
                                            content: [
                                                {
                                                    type: "text",
                                                    text: "Some stuff at the bottom",
                                                },
                                            ],
                                        },
                                    ],
                                },
                            ],
                        },
                    ],
                },
                rules,
                { prosemirrorDocWidth: 675 }
            ).asNode()
        ).toMatchSnapshot();
    });
});
