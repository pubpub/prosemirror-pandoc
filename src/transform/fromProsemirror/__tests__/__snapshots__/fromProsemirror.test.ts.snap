// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fromProsemirror converts a table 1`] = `
Object {
  "attr": Object {
    "classes": Array [],
    "identifier": "",
    "properties": Object {},
  },
  "bodies": Array [
    Object {
      "attr": Object {
        "classes": Array [],
        "identifier": "",
        "properties": Object {},
      },
      "bodyRows": Array [
        Object {
          "attr": Object {
            "classes": Array [],
            "identifier": "",
            "properties": Object {},
          },
          "cells": Array [
            Object {
              "alignment": "AlignDefault",
              "attr": Object {
                "classes": Array [],
                "identifier": "",
                "properties": Object {},
              },
              "colSpan": 1,
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "content": "Role",
                      "type": "Str",
                    },
                  ],
                  "type": "Para",
                },
              ],
              "rowSpan": 1,
              "type": "Cell",
            },
            Object {
              "alignment": "AlignDefault",
              "attr": Object {
                "classes": Array [],
                "identifier": "",
                "properties": Object {},
              },
              "colSpan": 1,
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "content": "Hero",
                      "type": "Str",
                    },
                  ],
                  "type": "Para",
                },
              ],
              "rowSpan": 1,
              "type": "Cell",
            },
            Object {
              "alignment": "AlignDefault",
              "attr": Object {
                "classes": Array [],
                "identifier": "",
                "properties": Object {},
              },
              "colSpan": 1,
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "content": "Villain",
                      "type": "Str",
                    },
                  ],
                  "type": "Para",
                },
              ],
              "rowSpan": 1,
              "type": "Cell",
            },
          ],
          "type": "Row",
        },
        Object {
          "attr": Object {
            "classes": Array [],
            "identifier": "",
            "properties": Object {},
          },
          "cells": Array [
            Object {
              "alignment": "AlignDefault",
              "attr": Object {
                "classes": Array [],
                "identifier": "",
                "properties": Object {},
              },
              "colSpan": 1,
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "content": "Weapon",
                      "type": "Str",
                    },
                  ],
                  "type": "Para",
                },
              ],
              "rowSpan": 1,
              "type": "Cell",
            },
            Object {
              "alignment": "AlignDefault",
              "attr": Object {
                "classes": Array [],
                "identifier": "",
                "properties": Object {},
              },
              "colSpan": 1,
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "content": "Power",
                      "type": "Str",
                    },
                    Object {
                      "type": "Space",
                    },
                    Object {
                      "content": "Sword",
                      "type": "Str",
                    },
                  ],
                  "type": "Para",
                },
              ],
              "rowSpan": 1,
              "type": "Cell",
            },
            Object {
              "alignment": "AlignDefault",
              "attr": Object {
                "classes": Array [],
                "identifier": "",
                "properties": Object {},
              },
              "colSpan": 1,
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "content": "Havoc",
                      "type": "Str",
                    },
                    Object {
                      "type": "Space",
                    },
                    Object {
                      "content": "Staff",
                      "type": "Str",
                    },
                  ],
                  "type": "Para",
                },
              ],
              "rowSpan": 1,
              "type": "Cell",
            },
          ],
          "type": "Row",
        },
        Object {
          "attr": Object {
            "classes": Array [],
            "identifier": "",
            "properties": Object {},
          },
          "cells": Array [
            Object {
              "alignment": "AlignDefault",
              "attr": Object {
                "classes": Array [],
                "identifier": "",
                "properties": Object {},
              },
              "colSpan": 1,
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "content": "Dark",
                      "type": "Str",
                    },
                    Object {
                      "type": "Space",
                    },
                    Object {
                      "content": "secret",
                      "type": "Str",
                    },
                  ],
                  "type": "Para",
                },
              ],
              "rowSpan": 1,
              "type": "Cell",
            },
            Object {
              "alignment": "AlignDefault",
              "attr": Object {
                "classes": Array [],
                "identifier": "",
                "properties": Object {},
              },
              "colSpan": 1,
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "content": "Expert",
                      "type": "Str",
                    },
                    Object {
                      "type": "Space",
                    },
                    Object {
                      "content": "florist",
                      "type": "Str",
                    },
                  ],
                  "type": "Para",
                },
              ],
              "rowSpan": 1,
              "type": "Cell",
            },
            Object {
              "alignment": "AlignDefault",
              "attr": Object {
                "classes": Array [],
                "identifier": "",
                "properties": Object {},
              },
              "colSpan": 1,
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "content": "Cries",
                      "type": "Str",
                    },
                    Object {
                      "type": "Space",
                    },
                    Object {
                      "content": "at",
                      "type": "Str",
                    },
                    Object {
                      "type": "Space",
                    },
                    Object {
                      "content": "romcoms",
                      "type": "Str",
                    },
                  ],
                  "type": "Para",
                },
              ],
              "rowSpan": 1,
              "type": "Cell",
            },
          ],
          "type": "Row",
        },
        Object {
          "attr": Object {
            "classes": Array [],
            "identifier": "",
            "properties": Object {},
          },
          "cells": Array [
            Object {
              "alignment": "AlignDefault",
              "attr": Object {
                "classes": Array [],
                "identifier": "",
                "properties": Object {},
              },
              "colSpan": 3,
              "content": Array [
                Object {
                  "content": Array [
                    Object {
                      "content": "Some",
                      "type": "Str",
                    },
                    Object {
                      "type": "Space",
                    },
                    Object {
                      "content": "stuff",
                      "type": "Str",
                    },
                    Object {
                      "type": "Space",
                    },
                    Object {
                      "content": "at",
                      "type": "Str",
                    },
                    Object {
                      "type": "Space",
                    },
                    Object {
                      "content": "the",
                      "type": "Str",
                    },
                    Object {
                      "type": "Space",
                    },
                    Object {
                      "content": "bottom",
                      "type": "Str",
                    },
                  ],
                  "type": "Para",
                },
              ],
              "rowSpan": 1,
              "type": "Cell",
            },
          ],
          "type": "Row",
        },
      ],
      "headRows": Array [],
      "rowHeadColumns": 1,
      "type": "TableBody",
    },
  ],
  "caption": Object {
    "content": Array [],
    "type": "Caption",
  },
  "colSpecs": Array [
    Object {
      "alignment": "AlignDefault",
      "type": "ColSpec",
      "width": 0.1985185185185185,
    },
    Object {
      "alignment": "AlignDefault",
      "type": "ColSpec",
      "width": 0.49333333333333335,
    },
    Object {
      "alignment": "AlignDefault",
      "type": "ColSpec",
      "width": 0.3985185185185185,
    },
  ],
  "foot": Object {
    "attr": Object {
      "classes": Array [],
      "identifier": "",
      "properties": Object {},
    },
    "rows": Array [],
    "type": "TableFoot",
  },
  "head": Object {
    "attr": Object {
      "classes": Array [],
      "identifier": "",
      "properties": Object {},
    },
    "rows": Array [
      Object {
        "attr": Object {
          "classes": Array [],
          "identifier": "",
          "properties": Object {},
        },
        "cells": Array [
          Object {
            "alignment": "AlignDefault",
            "attr": Object {
              "classes": Array [],
              "identifier": "",
              "properties": Object {},
            },
            "colSpan": 1,
            "content": Array [
              Object {
                "content": Array [],
                "type": "Para",
              },
            ],
            "rowSpan": 1,
            "type": "Cell",
          },
          Object {
            "alignment": "AlignDefault",
            "attr": Object {
              "classes": Array [],
              "identifier": "",
              "properties": Object {},
            },
            "colSpan": 1,
            "content": Array [
              Object {
                "content": Array [
                  Object {
                    "content": "He-Man",
                    "type": "Str",
                  },
                ],
                "type": "Para",
              },
            ],
            "rowSpan": 1,
            "type": "Cell",
          },
          Object {
            "alignment": "AlignDefault",
            "attr": Object {
              "classes": Array [],
              "identifier": "",
              "properties": Object {},
            },
            "colSpan": 1,
            "content": Array [
              Object {
                "content": Array [
                  Object {
                    "content": "Skeletor",
                    "type": "Str",
                  },
                ],
                "type": "Para",
              },
            ],
            "rowSpan": 1,
            "type": "Cell",
          },
        ],
        "type": "Row",
      },
    ],
    "type": "TableHead",
  },
  "type": "Table",
}
`;
